db/version_set.cc

```diff
@@ -4909,29 +4909,57 @@ uint64_t VersionSet::ApproximateSize(Version* v, const Slice& start,
     assert(files_brief.num_files > 0);
 
     // identify the file position for starting key
-    const uint64_t idx_start = FindFileInRange(
-        v->cfd_->internal_comparator(), files_brief, start,
-        /*start=*/0, static_cast<uint32_t>(files_brief.num_files - 1));
-    assert(idx_start < files_brief.num_files);
-
-    // scan all files from the starting position until the ending position
-    // inferred from the sorted order
-    for (uint64_t i = idx_start; i < files_brief.num_files; i++) {
-      uint64_t val;
-      val = ApproximateSize(v, files_brief.files[i], end, caller);
-      if (!val) {
-        // the files after this will not have the range
-        break;
+    const int idx_start =
+        FindFileInRange(v->cfd_->internal_comparator(), files_brief, start, 0,
+                        static_cast<uint32_t>(files_brief.num_files - 1));
+    assert(static_cast<size_t>(idx_start) < files_brief.num_files);
+    const int idx_end = FindFileInRange(
+        v->cfd_->internal_comparator(), files_brief, end, idx_start,
+        static_cast<uint32_t>(files_brief.num_files - 1));
+    assert(idx_end >= idx_start &&
+           static_cast<size_t>(idx_end) < files_brief.num_files);
+
+    // scan all files from the starting index to the ending index
+    // (inferred from the sorted order)
+
+    // first scan all the intermediate files, excluding first and last
+    for (int i = idx_start + 1; i < idx_end; ++i) {
+      uint64_t file_size = files_brief.files[i].fd.GetFileSize();
+      // The entire file falls into the range, so we can just take its size.
+      assert(file_size ==
+             ApproximateSize(v, files_brief.files[i], end, caller));
+      size += file_size;
+    }
+
+    // next scan the first and the last files (which may be the same file).
+    // we prefer to start with the last file, since it may become cheaper
+    // to estimate the size for the first file, if size grows sufficiently.
+    std::vector<int> first_last_files{idx_end};
+    if (idx_start != idx_end) {
+      // we need to estimate size for both files, only if they are different
+      first_last_files.push_back(idx_start);
+    }
+
+    for (int i : first_last_files) {
```

Comment:
It's good. I think it may be helpful to do it after calculating size of files of all levels.
And think it makes more sense to relax it only if sum of size of all the partial files is still much smaller than file size for full files.

Reply:
not sure i fully got this comment